File: examples/04_application/main.go
// Lean Code — BSD 3-Clause License — Vivian Voss, 2026
// Scope: Entry point. Process all orders into invoices.

package main

import (
	"fmt"
	"os"
	"strconv"

	"lean-code/billing"
	"lean-code/configuration"
	"lean-code/order"
	"lean-code/stock"
)

func main() {
	config := configuration.ReadConfigurationFromFile("config.txt")
	pricePerUnit, _ := strconv.ParseFloat(config.PricePerUnit, 64)

	orderFiles := order.FindOrderFilesInDirectory(config.OrderDirectory)

	for _, file := range orderFiles {
		o := order.ReadOrderFromFile(file)

		if !stock.CheckStockAvailability(o.Product, o.Quantity, config.StockFile) {
			fmt.Fprintf(os.Stderr, "out of stock: %s\n", o.Product)
			continue
		}

		invoice := billing.CreateInvoiceFromOrder(o.ID, o.Product, o.Quantity, pricePerUnit)
		billing.WriteInvoiceToFile(invoice, fmt.Sprintf("invoice_%d.txt", o.ID))
	}
}

File: examples/04_application/main.js
// Lean Code — BSD 3-Clause License — Vivian Voss, 2026
// Scope: Entry point. Process all orders into invoices.

import { readConfigurationFromFile } from "./configuration/configuration.js";
import { readOrderFromFile, findOrderFilesInDirectory } from "./order/order.js";
import { checkStockAvailability } from "./stock/stock.js";
import { createInvoiceFromOrder, writeInvoiceToFile } from "./billing/billing.js";

const configuration = readConfigurationFromFile("config.txt");
const pricePerUnit = parseFloat(configuration["price_per_unit"]);

const orderFiles = findOrderFilesInDirectory(configuration["order_directory"]);

for (const file of orderFiles) {
  const order = readOrderFromFile(file);

  if (!checkStockAvailability(order.product, order.quantity, configuration["stock_file"])) {
    process.stderr.write(`out of stock: ${order.product}\n`);
    continue;
  }

  const invoice = createInvoiceFromOrder(order, pricePerUnit);
  writeInvoiceToFile(invoice, `invoice_${order.id}.txt`);
}
